File: projects/pinstripe/packages/@pinstripe/core/lib/database/table.js

import { Base } from '../base.js';
import { Registrable } from '../registrable.js';
import { serviceFactory } from '../service_factory.js';
import { Inflector } from '../inflector.js';
import { Sql } from './sql.js';
import { MYSQL_COLUMN_TYPE_TO_TYPE_MAP, Column } from './column.js';
import { Union } from './union.js';

const COMPARISON_OPERATORS = {
    Eq: (column, value) => column.sql`${column} = ${value}`,
    Ne: (column, value) => column.sql`${column} != ${value}`,
    Lt: (column, value) => column.sql`${column} < ${value}`,
    Gt: (column, value) => column.sql`${column} > ${value}`,
    Le: (column, value) => column.sql`${column} <= ${value}`,
    Ge: (column, value) => column.sql`${column} >= ${value}`,
    BeginsWith: (column, value) => column.sql`${column} like concat(${value}, '%')`,
    EndsWith: (column, value) => column.sql`${column} like concat('%', ${value})`,
    Contains: (column, value) => column.sql`${column} like concat('%', ${value}, '%')`
};

const COMPARISON_OPERATOR_METHOD_PATTERN = new RegExp(`^(.+)(${Object.keys(COMPARISON_OPERATORS).join('|')})$`);

export const Table = Base.extend().define(dsl => dsl
    .include(Registrable)
    .tap(Class => {
        const register = Class.register;
        Class.define(dsl => dsl
            .classProps({
                register(name){
                    serviceFactory(name, ({ database }) => database[name]);
                    return register.call(this, name);
                }
            })
        );
    })
    .classProps({
        get relationships(){
            if(!this.hasOwnProperty('_relationships')){
                this._relationships = {};
            }
            return this._relationships;
        },

        toSql(){
            return Sql.escapeIdentifier(this.name);
        },

        ['dsl.hasMany'](name, options = {}){
            this.relationships[name] = {
                name,
                collectionName: name,
                fromKey: 'id',
                toKey: `${Inflector.singularize(this.name)}_id`,
                cascadeDelete: true,
                ...options
            };
            this['dsl.props']({
                [name](){
                    return this._join(name);
                }
            });
        },

        ['dsl.hasOne'](name, options = {}){
            this.relationships[name] = {
                name,
                collectionName: Inflector.pluralize(name),
                fromKey: 'id',
                toKey: `${name}_id`,
                cascadeDelete: true,
                ...options
            };
            this['dsl.props']({
                [name](){
                    return this._join(name);
                }
            });
        },

        ['dsl.belongsTo'](name, options = {}){
            this.relationships[name] = {
                name,
                collectionName: Inflector.pluralize(name),
                fromKey: `${name}_id`,
                toKey: 'id',
                cascadeDelete: false,
                ...options
            };
            this['dsl.props']({
                [name](){
                    return this._join(name);
                }
            });
        }
    })
    .props({
        initialize(database, joinParent){
            this._database = database;
            this._joinParent = joinParent;
            if(!this._joinParent){
                this._aliasCounters = {};
            }
            this._alias = this._generateAlias(this.constructor.name);

            if(!this._joinParent){
                this._fromSql = [this.sql`${this.constructor} as ${this}`];
                this._whereSql = [];
                this._orderBySql = [];
                this._limitSql = [];
            }
        },

        concat(collection){
            return new Union(this._database, [this, collection]);
        },

        get sql(){
            return Sql.fromTemplate(this);
        },

        toSql(){
            return Sql.escapeIdentifier(this._alias);
        },

        back(count = 1){
            let out = this;
            for(let i = 0; i < count; i++){
                out = out._joinParent;
            }
            return out;
        },

        where(...args){
            const whereSql = this._joinRoot._whereSql;
            if(whereSql.length){
                whereSql.push(' and ');
            }
            whereSql.push(this.sql(...args));
            return this;
        },

        orderBy(column, direction = 'asc'){
            if(!(column instanceof Column)){
                column = this.__getMissing(column);
            }
            const orderBySql = this._joinRoot._orderBySql;
            if(orderBySql.length){
                orderBySql.push(this.sql`, ${column} ${[direction == 'desc' ? 'desc' : 'asc']}`);
            } else {
                orderBySql.push(this.sql`${column} ${[direction == 'desc' ? 'desc' : 'asc']}`);
            }
            return this;
        },

        clearOrderBy(){
            this._joinRoot._orderBySql = [];
            return this;
        },

        paginate(page = 1, pageSize = 10){
            this._joinRoot._limitSql = [this.sql`${(page - 1) * pageSize}, ${pageSize}`];
            return this;
        },

        clearPagination(){
            this._joinRoot._limitSql = [];
            return this;
        },

        all(options = {}){
            return this._database.run`${this._generateSelectSql(options)}`;
        },

        async forEach(fn){
            const rows = await this.all();
            for(let i = 0; i < rows.length; i++){
                const row = rows[i];
                await fn.call(row, row, i);
            }
            return this;
        },

        async first(options = {}){
            await this.paginate(1, 1);
            return (await this.all(options)).pop();
        },

        async count(options = {}){
            return Object.values(await this.first({select: 'count(*)', ...options})).pop();
        },

        explain(options = {}){
            return this._generateSelectSql(options);
        },

        async insert(...args){
            const { Row } = await import('./row.js');
            await Row.create(Inflector.singularize(this.constructor.name), this._database).update(...args);
            return this;
        },

        async update(...args){
            await this._database.transaction(async () => {
                await this.forEach(row => row.update(...args));
            });
            return this;
        },

        async delete(){
            await this._database.transaction(async () => {
                await this.forEach(row => row.delete());
            });
            return this;
        },

        async columns(){
            const out = {};
            if(await this.exists()){
                const rows = await this._database.run`describe ${this.constructor}`;
                rows.forEach(row => {
                    const name = row['Field'];
                    let type;
                    if(name == 'id'){
                        type = 'primary_key'
                    } else {
                        type = MYSQL_COLUMN_TYPE_TO_TYPE_MAP[row['Type']] || 'string';
                    }
                    out[name] = new Column(this, name, type);
                });
            }
            return out;
        },

        async exists(){
            return Object.keys(await this._database.tables()).includes(this.constructor.name);
        },

        async create(){
            await this._database.create();
            if(!await this.exists()){
                await this._database.run`
                    create table ${this.constructor}(
                        id int(11) unsigned auto_increment primary key
                    )
                `;
            }
        },

        async drop(){
            if(await this.exists()){
                await this._database.run`drop table ${this.constructor}`;
            }
        },

            
        async addColumn(name, ...args){
            return (await this[name]).create(...args);
        },

        async removeColumn(name){
            return (await this[name]).drop();
        },

        async __getMissing(name){
            const columns = await this.columns();
            const matches = name.match(COMPARISON_OPERATOR_METHOD_PATTERN);
            if(matches){
                const column = columns[matches[1]];
                if(column){
                    return (value) => this.where`${COMPARISON_OPERATORS[matches[2]](column, value)}`;
                } else {
                    return () => this.where`1 = 2`;
                }
            }
            return columns[name] || new Column(this, name);
        },

        _join(relationshipName){
            const relationship = this.constructor.relationships[relationshipName]
            if(Union.classes[relationship.collectionName]){
                return this._joinToUnion(relationship);
            }   
            return this._joinToTable(relationship);
        },

        _joinToUnion(relationship){
            const unionClass = Union.classes[relationship.collectionName]
            const tables = unionClass.tableClasses.map(tableClass => {
                const out = new tableClass(this._database, this);
                const joinRoot = out._joinRoot;
                const fromSql = joinRoot._fromSql;
                const whereSql = joinRoot._whereSql;
                fromSql.push(out.sql`, ${out.constructor} as ${out}`);
                whereSql.push(out.sql`${[whereSql.length ? ' and ' : '']}${this[relationship.fromKey]} = ${this[relationship.toKey]}`);
                let matches;
                if(matches = relationship.fromKey.match(/^(.+)_id$/)){
                    whereSql.push(out.sql`and ${this[`${matches[1]}Type`]} = ${Inflector.singularize(out.constructor.name)}`);
                } else if(matches = relationship.to_key.match(/\A(.+)_id\z/)){
                    whereSql.push(out.sql`and ${this[`${matches[1]}Type`]} = ${Inflector.singularize(out.constructor.name)}`);
                }
                return out;
            });
            return new unionClass(this._database, tables);
        },

        _joinToTable(relationship){
            const out = Table.create(relationship.collectionName, this._database, this);
            const joinRoot = out._joinRoot;
            const fromSql = joinRoot._fromSql;
            const whereSql = joinRoot._whereSql;
            fromSql.push(out.sql`, ${out.constructor} as ${out}`);
            whereSql.push(out.sql`${[whereSql.length ? ' and ' : '']}${this[relationship.fromKey]} = ${this[relationship.toKey]}`);
            return out;
        },

        async _generateSelectSql(options = {}){
            const columns = await this.columns();
            options = {
                columnNames: Object.keys(columns),
                ...options
            };

            const out = ['select '];

            if(options.select){
                out.push(options.select);
            } else {
                for(let i in options.columnNames){
                    const columnName = options.columnNames[i];
                    const column = await this.__getMissing(columnName);
                    if(await column.exists()){
                        out.push(this.sql`${column} as ${Sql.escapeIdentifier(columnName)}, `);
                    } else {
                        out.push(this.sql`null as ${Sql.escapeIdentifier(columnName)}, `);
                    }
                }
                out.push(this.sql`${Inflector.singularize(this.constructor.name)} as \`_type\``);
            }

            const joinRoot = this._joinRoot;
            
            if(options.from !== undefined){
                if(options.from){
                    out.push(this.sql` from ${[options.from]}`);
                }
            } else {
                out.push(this.sql` from ${joinRoot._fromSql}`);
            }

            if (options.where !== undefined){
                if(options.where){
                    out.push(this.sql` where ${[options.where]}`);
                }
            } else if(joinRoot._whereSql.length) {
                out.push(this.sql` where ${joinRoot._whereSql}`);
            }

            if(options.orderBy !== undefined){
                if(options.orderBy){
                    out.push(this.sql` order by ${[options.orderBy]}`);
                }
            } else if(joinRoot._orderBySql.length){
                out.push(this.sql` order by ${joinRoot._orderBySql}`);
            }

            if(options.limit !== undefined){
                if(options.limit){
                    out.push(this.sql` limit ${[options.limit]}`);
                }
            } else if(joinRoot._limitSql.length) {
                out.push(this.sql` limit ${joinRoot._limitSql}`);
            }

            return this.sql`${out}`;
        },

        _generateAlias(name){
            const aliasCounters = this._joinRoot._aliasCounters;

            if(aliasCounters[name] === undefined){
                aliasCounters[name] = 0;
            }
            aliasCounters[name]++;

            if(aliasCounters[name] == 1){
                return name;
            }

            return `${name}${aliasCounters[name]}`;
        },

        get _joinRoot(){
            if(!this.__joinRoot){
                const out = this;
                while(out._joinParent){
                    out = out._joinParent;
                }
                this.__joinRoot = out;
            }
            return this.__joinRoot;
        }
    })
);

File: projects/slick/lib/slick/database/table.rb


require "slick/registry"
require "slick/database/adaptable"
require "slick/helper"

class Slick::Database::Table

    class << self

        include Slick::Registry
        include Slick::Database::Adaptable

        def register(name, *args, &block)
            out = super
            Slick::Helper.register(name).define_method "call" do |*args, &block|
                database.send(self.class.name, *args, &block)
            end
            out
        end

        def relationships
            @relationships ||= {}
        end

        def has_many(name, options = {})
            name = name.to_s

            relationships[name] = {
                name: name.to_s,
                collection_name: name.to_s,
                from_key: "id",
                to_key: "#{self.name.singularize}_id",
                cascade_delete: true
            }
            
            relationships[name].merge!(options)

            relationships[name][:cascade_delete] = false if !relationships[name][:through].nil?

            class_eval("def #{name}; _join('#{name}'); end")
        end

        def has_one(name, options = {})
            name = name.to_s

            relationships[name] = {
                name: name.to_s,
                collection_name: name.pluralize,
                from_key: "id",
                to_key: "#{self.name.singularize}_id",
                cascade_delete: true
            }
            relationships[name].merge!(options)
            class_eval("def #{name}; _join('#{name}'); end")
        end

        def belongs_to(name, options = {})
            name = name.to_s

            relationships[name] = {
                name: name.to_s,
                collection_name: name.pluralize,
                from_key: "#{name}_id",
                to_key: "id",
                cascade_delete: false
            }
            relationships[name].merge!(options)
            class_eval("def #{name}; _join('#{name}'); end")
        end

    end

    _define_adapter_methods :table

    deligate_to_adapter :columns, :create, :drop, :method_missing, :_generate_select_sql

    attr_reader :database

    def initialize(database, join_parent = nil)
        @database = database
        @join_parent = join_parent
        @alias_counters = {} if !@join_parent
        @alias = _generate_alias(self.class.name)

        if !@join_parent
            @from_sql = ['? as ?', self.class.name.to_sym, @alias.to_sym]
            @where_sql = []
            @order_by_sql = []
            @skip_count = 0
        end
    end

    def +(collection)
        Slick::Database::Union.new(@database, [clone, collection])
    end

    def clone
        out = super
        out.instance_eval do
            instance_variables.each do |name|
                if name.match(/\A@_/)
                    remove_instance_variable(name)
                elsif name != :@database
                    instance_variable_set(name, instance_variable_get(name).clone)
                end
            end
        end
        out
    end

    def back(count = 1)
        out = self
        count.times{ out = out.instance_variable_get(:@join_parent) }
        out
    end

    def where(*predicate, &block)
        if block
            join_path = []
            current = self
            block.arity.times do
                join_path << current
                current = current.instance_eval{ @join_parent } if current
            end
            predicate << block.call(*join_path)
        end
        out = clone
        where_sql = out.send(:_join_root).instance_variable_get(:@where_sql)
        if where_sql.count > 0
            where_sql << ['and', predicate]
        else
            where_sql << predicate
        end
        out
    end

    def order_by(column, direction = 'asc')
        out = clone
        column = out.send(column.to_s) if !column.kind_of?(Slick::Database::Column)
        order_by_sql = out.send(:_join_root).instance_variable_get(:@order_by_sql)
        if order_by_sql.count > 0
            order_by_sql << [", ? #{direction.to_s == 'desc' ? 'desc' : 'asc'}", column]
        else
            order_by_sql << ["? #{direction.to_s == 'desc' ? 'desc' : 'asc'}", column]
        end
        out
    end

    def clear_order_by
        out = clone
        out.send(:_join_root).instance_variable_set(:@order_by_sql, [])
        out
    end

    def paginate(page = 1, page_size = 10)
        out = clone
        out.send(:_join_root).instance_eval do
            @page = page.to_i
            @page_size = page_size.to_i
        end
        out
    end

    def clear_pagination
        out = clone
        out.send(:_join_root).instance_eval do
            @page = nil
            @page_size = nil
        end
        out
    end

    def skip(skip_count)
        out = clone
        out.send(:_join_root).instance_eval do
            @skip_count = skip_count
        end
        out
    end

    def all(options = {})
        @database.run(_generate_select_sql(options), options)
    end

    def each(options = {}, &block)
        @database.run(_generate_select_sql(options), options).each(&block)
    end

    def first(options = {})
        clone.paginate(1, 1).all(options).pop
    end

    def [](id)
        clone.id_eq(id).first
    end

    def count(options = {})
        first(options.merge(select: ['count(distinct ?)', id])).values.pop
    end

    def explain(options = {})
        @database.send(:_prepare, _generate_select_sql(options))
    end

    def inspect
        all
    end

    def insert(fields = {}, &block)
        Slick::Database::Row.create(self.class.name.singularize, @database).update(fields, &block)
    end

    def update(fields = {}, &block)
        @database.transaction do
            each{|row| row.update(fields, &block)}
        end
        self
    end

    def delete
        @database.transaction do
            each{|row| row.delete}
        end
        self
    end
    
    def exist?
        @database.tables.keys.include?(self.class.name)
    end
    
    def add_column(name, *args)
        send(name).create(*args)
    end

    def remove_column(name)
        send(name).drop
    end

    COLUMN_TYPE_TO_FORM_FIELD_TYPE_MAP = {
        "primary_key" => "hidden",
        "alternate_key" => "hidden",
        "foreign_key" => "hidden",
        "binary" => "file",
        "boolean" => "checkbox",
        "date" => "date",
        "time" => "datetime-local",
        "decimal" => "number",
        "float" => "number",
        "integer" => "number",
        "string" => "text",
        "text" => "textarea",
    }

    def to_form_adapter
        Class.new do
            def initialize(table)
                @table = table
            end

            def title
                "Add #{@table.class.name.singularize}"
            end

            def fields
                @fields ||= begin
                    out = [];
                    @table.columns.values.each do |column|
                        out << {
                            name: column.name,
                            type: COLUMN_TYPE_TO_FORM_FIELD_TYPE_MAP[column.type]
                        }
                    end
                    out
                end
                
            end

            def submit_title
                title
            end

            def cancel_title
                "Cancel"
            end

            def unsaved_changes_confirm
                "There are unsaved changes are you sure you want to close?"
            end

            def submit(values, &block)
                block.call(@table.insert(values))
            end
        end.new(self)
    end

    private

    def _join(relationship_name)
        relationship = self.class.relationships[relationship_name]

        if relationship[:through]
            out = self
            relationship[:through].each do |relationship|
                out = out.send(relationship)
            end
            out
        elsif Slick::Database::Union.registered_classes[relationship[:collection_name]]
            _join_to_union(relationship)
        else
            _join_to_table(relationship)
        end
    end

    def _join_to_union(relationship)
        union_class = Slick::Database::Union.registered_classes[relationship[:collection_name]]
        tables = union_class.table_classes.map do |table_class|
            join_parent = clone
            out = table_class.new(@database, join_parent)
            out.send(:_join_root).instance_eval do
                @from_sql << [', ? as ?', out.class.name.to_sym, out.instance_variable_get(:@alias).to_sym]
                @where_sql << ["#{@where_sql.count > 0 ? 'and ' : ''}? = ?", join_parent.send(relationship[:from_key]), out.send(relationship[:to_key])]
                if matches = relationship[:from_key].match(/\A(.+)_id\z/)
                    @where_sql << ['and ? = ?', join_parent.send("#{matches[1]}_type"), out.class.name.singularize]
                elsif matches = relationship[:to_key].match(/\A(.+)_id\z/)
                    @where_sql << ['and ? = ?', out.send("#{matches[1]}_type"), join_parent.class.name.singularize]
                end
            end
            out
        end
        union_class.new(@database, tables)
    end

    def _join_to_table(relationship)
        join_parent = clone
        out = Slick::Database::Table.create(relationship[:collection_name], @database, join_parent)
        out.send(:_join_root).instance_eval do
            @from_sql << [', ? as ?', out.class.name.to_sym, out.instance_variable_get(:@alias).to_sym]
            @where_sql << ["#{@where_sql.count > 0 ? 'and ' : ''}? = ?", join_parent.send(relationship[:from_key]), out.send(relationship[:to_key])]
        end
        out
    end

    def _generate_alias(name)
        alias_counters = _join_root.instance_variable_get(:@alias_counters)

        name = name.to_s
        alias_counters[name] = 0 if alias_counters[name].nil?
        alias_counters[name] += 1

        if alias_counters[name] == 1
            name
        else
            "#{name}#{alias_counters[name]}"
        end
    end

    def _join_root
        @_join_root ||= begin
            out = self
            while join_parent = out.instance_variable_get(:@join_parent)
                out = join_parent
            end
            out
        end
    end

    def _adapter
        @_adapter ||= database.send(:_adapter)
    end
end
